tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/RawGradientAdapter.java

```diff
@@ -0,0 +1,75 @@
+/*
+ Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================
+
+*/
+package org.tensorflow.op;
+
+import static org.tensorflow.internal.c_api.global.tensorflow.StatusFromTF_Status;
+
+import java.util.List;
+import org.bytedeco.javacpp.PointerScope;
+import org.tensorflow.BaseGradientAdapter;
+import org.tensorflow.Graph;
+import org.tensorflow.GraphOperation;
+import org.tensorflow.Operand;
+import org.tensorflow.Output;
+import org.tensorflow.internal.c_api.NativeOperation;
+import org.tensorflow.internal.c_api.NativeOutputVector;
+import org.tensorflow.internal.c_api.NativeStatus;
+import org.tensorflow.internal.c_api.TF_Scope;
+import org.tensorflow.internal.c_api.TF_Status;
+
+/** A native adapter for {@link RawCustomGradient}. */
+final class RawGradientAdapter extends BaseGradientAdapter {
+
+  private final RawCustomGradient gradient;
+
+  RawGradientAdapter(RawCustomGradient gradient) {
+    super();
+    this.gradient = gradient;
+  }
+
+  @Override
+  public NativeStatus call(
+      TF_Scope scope,
+      NativeOperation op,
+      NativeOutputVector grad_inputs,
+      NativeOutputVector grad_outputs) {
+    try (PointerScope pointerScope = new PointerScope()) {
+      Graph g = Graph.findGraphForPointer(scope.graph());
+      if (g == null) {
+        throw new IllegalStateException("No graph found for native gradient scope.");
+      }
+
+      Scope nativeScope = new GradientScope(scope, g);
```

Comment:
This scope is unnamed. Can we put the op name in?

Reply:
I'm unsure whether the native scope it's created from is already named, will check.

Reply:
Can't get things to run locally at the moment, I'll check the CI logs once it runs all the way through.

Reply:
It looks like the scopes aren't named in the native code.  So I could name it here, but then all the non-custom gradients will just be in `gradients/` while the custom gradients will be in `gradients/opName/`.  Imo matching non-custom gradients is more important here.

Reply:
I think the gradients output by the C API aren't very well named as it's very hard to figure out what's going on from tensorboard. So in this case I think we should add more information in things we can control with the aim of fixing the C API later.

Reply:
:+1: